File: audits/clusterprofiler/verify/gsea_exact_small.py
"""Exact enumeration of the gene-set permutation null for the tiny designs of gsea_exact_small.R.
Usage: python gsea_exact_small.py <outdir>"""
import json, sys, os
import numpy as np, pandas as pd
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from ref_gsea import exact_null, gsea_es, gsea_pvalue_from_null

out = sys.argv[1]
eng = pd.read_csv(os.path.join(out, "engines.tsv"), sep="\t")
nulls = {}
rows = []
for _, r in eng.iterrows():
    st = json.load(open(os.path.join(out, f"{r.design}_stats.json")))["stats"]
    s = pd.Series({k: v[0] if isinstance(v, list) else v for k, v in st.items()})
    sv = s.values; genes = np.array(s.index)
    key = (r.design, int(r.k))
    if key not in nulls:
        nulls[key] = exact_null(sv, int(r.k))
    null = nulls[key]
    hits = np.isin(genes, r.members.split("/"))
    ref = gsea_es(sv, hits)
    ex = gsea_pvalue_from_null(ref["ES"], null)
    rows.append(dict(design=r.design, set=r.set, ES_ref=ref["ES"], ES_fgsea=r.fgsea_ES, ES_enrichit=r.enrichit_ES,
                     p_exact_cond=ex["p_cond"], p_exact_uncond=ex["p_uncond"], NES_exact=ex["NES"],
                     fgsea_simple=r.fgsea_simple_p, fgsea_ml=r.fgsea_ml_p_mean, fgsea_ml_sd=r.fgsea_ml_p_sd,
                     enrichit_ml=r.enrichit_ml_p_mean, enrichit_ml_sd=r.enrichit_ml_p_sd,
                     enrichit_sample=r.enrichit_sample_p, enrichit_permute=r.enrichit_permute_p, enrichit_adaptive=r.enrichit_adaptive_p,
                     dose_perm=r.get("dose_perm_p", np.nan), NES_fgsea=r.fgsea_NES, NES_enrichit_sample=r.enrichit_sample_NES,
                     NES_dose=r.get("dose_perm_NES", np.nan), n_subsets=null.size))
df = pd.DataFrame(rows)
pd.set_option("display.width", 250)
print("ES: max |fgsea - ref| = %.2e, max |enrichit - ref| = %.2e" % ((df.ES_fgsea - df.ES_ref).abs().max(), (df.ES_enrichit - df.ES_ref).abs().max()))
print("\nExact conditional p = P(ES* >= ES | same sign)  vs engines  (multilevel: mean over 40 runs)")
cols = ["design", "set", "ES_ref", "n_subsets", "p_exact_cond", "p_exact_uncond", "fgsea_simple", "fgsea_ml", "enrichit_ml", "enrichit_sample", "enrichit_permute", "enrichit_adaptive", "dose_perm"]
print(df[cols].to_string(index=False, float_format=lambda x: f"{x:.4g}"))
print("\nratio engine / exact conditional p:")
for c in ["fgsea_simple", "fgsea_ml", "enrichit_ml", "enrichit_sample", "enrichit_permute", "enrichit_adaptive", "dose_perm"]:
    ratio = df[c] / df.p_exact_cond
    print(f"   {c:18s} median {ratio.median():.3f}  min {ratio.min():.3f}  max {ratio.max():.3f}")
print("\nratio enrichit sample / exact UNconditional p: median %.3f (min %.3f, max %.3f)" % ((df.enrichit_sample / df.p_exact_uncond).median(), (df.enrichit_sample / df.p_exact_uncond).min(), (df.enrichit_sample / df.p_exact_uncond).max()))
print("\nNES vs exact ES/mean(same-sign null ES):")
print(df[["design", "set", "NES_exact", "NES_fgsea", "NES_enrichit_sample", "NES_dose"]].to_string(index=False, float_format=lambda x: f"{x:.4f}"))

File: audits/clusterprofiler/verify/gsea_exact_small.R
## GSEA p-values against an exact enumeration of the gene-set permutation null on tiny lists.
## For each (N, k) design and each hand-picked gene set the engines report ES and p; the Python
## side enumerates all C(N, k) subsets and computes P(ES* >= ES | ES* >= 0) (the GSEA convention,
## which fgsea and DOSE document) and the unconditional P(ES* >= ES).
## Usage: R_LIBS_USER=<lib-legacy>:<lib> Rscript gsea_exact_small.R <outdir>
args <- commandArgs(TRUE); outdir <- if (length(args)) args[1] else tempdir()
dir.create(outdir, showWarnings = FALSE, recursive = TRUE)
suppressPackageStartupMessages({library(fgsea); library(enrichit); library(jsonlite)})
cat("fgsea", as.character(packageVersion("fgsea")), " enrichit", as.character(packageVersion("enrichit")), "\n")
has_dose <- requireNamespace("DOSE", quietly = TRUE) && packageVersion("DOSE") < "4.0.0"
if (has_dose) cat("DOSE", as.character(packageVersion("DOSE")), "\n")

designs <- list(
    list(name = "N24k3", N = 24, sets = list(top = 1:3, top_mid = c(1, 2, 12), mid = c(8, 12, 16), bottom = 22:24, mixed = c(2, 13, 23))),
    list(name = "N40k4", N = 40, sets = list(top = 1:4, top2 = c(1, 2, 3, 9), mid = c(15, 18, 22, 27), bottom = 37:40, mixed = c(1, 2, 39, 40))),
    list(name = "N30k5", N = 30, sets = list(top = 1:5, top2 = c(1, 2, 3, 4, 14), bottom = 26:30, mixed = c(1, 2, 15, 29, 30)))
)
REP <- 40
res <- list()
for (dz in designs) {
    set.seed(100 + dz$N)
    stats <- sort(rnorm(dz$N), decreasing = TRUE); names(stats) <- sprintf("g%02d", seq_len(dz$N))
    write_json(list(stats = as.list(stats)), file.path(outdir, paste0(dz$name, "_stats.json")), digits = NA)
    gsets <- lapply(dz$sets, function(ix) names(stats)[ix])
    k <- length(gsets[[1]])
    if (has_dose) {
        ## legacy DOSE::geneSet_filter() uses sapply(), which turns equal-sized sets into a matrix and then
        ## fails ("arguments imply differing number of rows"); pad the GSON with one set of another size
        gpad <- c(gsets, list(PAD = names(stats)[seq_len(k + 2)]))
        gs <- gson::gson(gsid2gene = data.frame(gsid = rep(names(gpad), lengths(gpad)), gene = unlist(gpad)),
                         gsid2name = data.frame(gsid = names(gpad), name = names(gpad)),
                         species = "s", gsname = "s", version = "0", accessed_date = "2026-09-08", keytype = "k")
        GI <- get("GSEA_internal", envir = asNamespace("DOSE"))
    }
    for (sn in names(gsets)) {
        one <- gsets[sn]
        set.seed(1); fs <- fgseaSimple(one, stats, nperm = 1e5, minSize = 1, maxSize = dz$N - 1, BPPARAM = BiocParallel::SerialParam())
        fml <- replicate(REP, { fgseaMultilevel(one, stats, minSize = 1, maxSize = dz$N - 1, eps = 0, BPPARAM = BiocParallel::SerialParam())$pval })
        eml <- replicate(REP, suppressWarnings(enrichit::gsea(stats, one, minGSSize = 1, maxGSSize = dz$N - 1, method = "multilevel", eps = 0, verbose = FALSE))$pvalue)
        esa <- suppressWarnings(enrichit::gsea(stats, one, minGSSize = 1, maxGSSize = dz$N - 1, method = "sample", nPerm = 2e5, seed = 5, verbose = FALSE))
        epe <- suppressWarnings(enrichit::gsea(stats, one, minGSSize = 1, maxGSSize = dz$N - 1, method = "permute", nPerm = 5e4, seed = 6, verbose = FALSE))
        ead <- suppressWarnings(enrichit::gsea(stats, one, minGSSize = 1, maxGSSize = dz$N - 1, method = "sample", adaptive = TRUE, minPerm = 1000, maxPerm = 2e5, seed = 7, verbose = FALSE))
        row <- list(design = dz$name, set = sn, N = dz$N, k = k, members = paste(one[[1]], collapse = "/"),
                    fgsea_ES = fs$ES, fgsea_simple_p = fs$pval, fgsea_nMoreExtreme = fs$nMoreExtreme, fgsea_NES = fs$NES,
                    fgsea_ml_p_mean = mean(fml), fgsea_ml_p_sd = sd(fml),
                    enrichit_ES = esa$enrichmentScore, enrichit_sample_p = esa$pvalue, enrichit_sample_NES = esa$NES,
                    enrichit_permute_p = epe$pvalue, enrichit_adaptive_p = ead$pvalue, enrichit_adaptive_nPerm = ead$nPerm,
                    enrichit_ml_p_mean = mean(eml), enrichit_ml_p_sd = sd(eml))
        if (has_dose) {
            set.seed(9)
            dd <- GI(geneList = stats, exponent = 1, minGSSize = 1, maxGSSize = dz$N - 1, eps = 0, pvalueCutoff = 1,
                     pAdjustMethod = "BH", verbose = FALSE, USER_DATA = gs, by = "DOSE", nPerm = 1e4)
            r <- dd@result; row$dose_perm_p <- r[sn, "pvalue"]; row$dose_perm_NES <- r[sn, "NES"]
        }
        res[[length(res) + 1]] <- as.data.frame(row)
        cat(sprintf("%s %-8s ES=%+.4f fgseaSimple p=%.4g  fgseaML %.4g±%.2g  enrichit sample %.4g  permute %.4g  adaptive %.4g (n=%d)  enrichit ML %.4g±%.2g%s\n",
                    dz$name, sn, fs$ES, fs$pval, mean(fml), sd(fml), esa$pvalue, epe$pvalue, ead$pvalue, ead$nPerm, mean(eml), sd(eml),
                    if (has_dose) sprintf("  DOSE perm %.4g", row$dose_perm_p) else ""))
    }
}
out <- do.call(rbind, res)
write.table(out, file.path(outdir, "engines.tsv"), sep = "\t", quote = FALSE, row.names = FALSE)
cat("done\n")
